runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -1999,56 +2005,79 @@ public Iterable<Entry<K, V>> read() {
-                        (k, v) -> {
-                          if (v == null) v = new ConcatIterables<>();
-                          v.extendWith(entry.getValue());
-                          keyState.existence = KeyExistence.KNOWN_EXIST;
-                          return v;
-                        });
-                  } catch (IOException e) {
-                    throw new RuntimeException(e);
-                  }
-                });
+            // If a key returned by windmill is known to be no longer exist or is completely cached
+            // locally, we can safely ignore the content of this key in windmill. Thus, we filter
+            // entries to filteredEntries which contains only keys that are known to exist and not
+            // fully cached.
+            Iterable<Entry<Object, Iterable<V>>> filteredEntries =
+                Iterables.filter(
+                    Iterables.transform(
+                        entries,
+                        entry -> {
+                          try {
+                            final K key = keyCoder.decode(entry.getKey().newInput());
+                            final Object structuralKey = keyCoder.structuralValue(key);
+                            KeyState keyState =
+                                keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
+                            if (keyState.existence == KeyExistence.UNKNOWN_EXISTENCE) {
+                              keyState.existence = KeyExistence.KNOWN_EXIST;
+                            }
+                            return new AbstractMap.SimpleEntry<>(structuralKey, entry.getValue());
+                          } catch (IOException e) {
+                            throw new RuntimeException(e);
+                          }
+                        }),
+                    entry -> {
+                      KeyState keyState = keyStateMap.get(entry.getKey());
+                      return keyState.existence != KeyExistence.KNOWN_NONEXISTENT
+                          && !(keyState.existence == KeyExistence.KNOWN_EXIST
+                              && keyState.valuesCached);
+                    });
+
             if (entries instanceof Weighted) {
               // This is a known amount of data, cache them all.
-              entryMap.forEach(
-                  (structuralKey, values) -> {
+              filteredEntries.forEach(
+                  entry -> {
+                    final Object structuralKey = entry.getKey();
                     KeyState keyState = keyStateMap.get(structuralKey);
-                    if (!keyState.valuesCached) {
-                      keyState.values.extendWith(values);
-                      keyState.valuesCached = true;
-                    }
+                    keyState.existence = KeyExistence.KNOWN_EXIST;
+                    keyState.values.extendWith(entry.getValue());
+                    // We can't set keyState.valuesCached to true here, because there may be more
```

Comment:
I think the weighted case could be simpler by just iterating over entries, merging with keyStateMap and then returning keyStateMap

I think the transform/filter above could be moved to just the lazy loading case.   I think you could end up with just a single lookup in keyStateMap for each entry.